core_backend/app/question_answer/routers.py

```diff
@@ -307,6 +320,39 @@ async def search_base(
     return response
 
 
+def is_unable_to_generate_response(response: QueryResponse) -> bool:
+    """
+    Check if the response is of type QueryResponseError and caused
+    by low alignment score.
+    """
+    return (
+        isinstance(response, QueryResponseError)
+        and response.error_type == ErrorType.ALIGNMENT_TOO_LOW
+    )
+
+
+@backoff.on_predicate(
+    backoff.expo,
+    max_tries=int(ALIGN_SCORE_N_RETRIES),
+    predicate=is_unable_to_generate_response,
+)
+async def retry_search(
```

Comment:
Decided to try the whole search again because the failure can come from any guardrail, so instead of rerunning only llm generation it's best to run everything again.

Reply:
I see how tricky it can be to just retry response generation + align score in the current setup.. 

How about we break up the `search_base` function into 
1. `search_pipeline`, with all the input check (__before) guardrails
2. `generation_pipeline` with generation + all the output check (__after) guardrails?

Like the following:
```python
@identify_language__before
@classify_safety__before
@translate_question__before
@paraphrase_question__before
async def search_pipeline(  
    query_refined: QueryRefined,
    response: QueryResponse,
    user_id: int,
    n_similar: int,
    asession: AsyncSession,
    exclude_archived: bool = True,
) -> QueryResponse | QueryResponseError:
    # always do the embeddings search even if some guardrails have failed
    metadata = create_langfuse_metadata(query_id=response.query_id, user_id=user_id)
    search_results = await get_similar_content_async(
        user_id=user_id,
        # use latest version of the text
        question=query_refined.query_text,
        n_similar=n_similar,
        asession=asession,
        metadata=metadata,
        exclude_archived=exclude_archived,
    )
    response.search_results = search_results

    return response


@generate_tts__after
@backoff.on_predicate(
    backoff.expo,
    max_tries=int(ALIGN_SCORE_N_RETRIES),
    predicate=is_unable_to_generate_response,
)  # <--- so that we don't run generate_tts unnecessarily.
@check_align_score__after
@generate_llm_response__after
async def generation_pipeline(
    query_refined: QueryRefined,
    response: QueryResponse,
    user_id: int,
    n_similar: int,
    asession: AsyncSession,
    exclude_archived: bool = True,
) -> QueryResponse | QueryResponseError:
    # Maybe put the `generate_llm_response__after` code here?
    return response
```

We'll probably have to change how we handle `generate_llm_response` inside `generate_llmresponse__after`.. Or as I mention above, just move the `generate_llm_response__after` code into `generation_pipeline.

Reply:
Thanks @suzinyou ,
1) What I meant is that, redoing each step might actually improve the alignScore for example:
- A new paraphased text can improve the LLM response and thus improve alignScore
- A new translation can improve the LLM response and so on 
That is the reason one I think it is beneficial to run the whole pipeline again. 
2) I do agree that rerunning search is wasteful but the whole reason why we are using decorators is to unify search and llm_response. Wouldn't what you suggest bring us backwards? I think this is worth discussing in a tech session before I actually implement the changes.

Reply:
1. Taking this to slack
2. I think it's actually quite opaque how we handle generate_llm_response inside a decorator, so it would be good to bring it out to the router logic. It is easier to understand, and we still keep the checks and transformations (guardrails) as decorators.